src/atomate2/abinit/jobs/response.py

```
@@ -0,0 +1,297 @@
+"""Response function jobs for running ABINIT calculations."""
+
+from __future__ import annotations
+
+import itertools
+import logging
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, ClassVar
+
+from abipy.flowtk.events import (
+    AbinitCriticalWarning,
+    NscfConvergenceWarning,
+    ScfConvergenceWarning,
+)
+from jobflow import Flow, Job, Response, job
+
+from atomate2.abinit.jobs.base import BaseAbinitMaker, abinit_job
+from atomate2.abinit.powerups import update_user_abinit_settings
+from atomate2.abinit.sets.response import (
+    DdeSetGenerator,
+    DdkSetGenerator,
+    DteSetGenerator,
+)
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from abipy.abio.inputs import AbinitInput
+    from pymatgen.core.structure import Structure
+
+    from atomate2.abinit.sets.base import AbinitInputGenerator
+    from atomate2.abinit.utils.history import JobHistory
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "ResponseMaker",
+    "DdkMaker",
+    "DdeMaker",
+    "DteMaker",
+    "generate_ddk_perts",
+    "generate_dde_perts",
+    "generate_dte_perts",
+    "run_rf",
+]
+
+
+@dataclass
+class ResponseMaker(BaseAbinitMaker):
+    """Maker for a Response Function ABINIT calculation job.
+
+    Parameters
+    ----------
+    calc_type : str
+        The type of RF.
+    name : str
+        The job name.
+    """
+
+    calc_type: str | None = None
+    name: str = "RF calculation"
+    input_set_generator: AbinitInputGenerator | None = None
+
+    CRITICAL_EVENTS: ClassVar[Sequence[AbinitCriticalWarning]] = (
+        ScfConvergenceWarning,
+    )
+
+    @abinit_job
+    def make(
+        self,
+        structure: Structure | None = None,
+        prev_outputs: str | list[str] | None = None,
+        restart_from: str | list[str] | None = None,
+        history: JobHistory | None = None,
+        perturbation: dict | None = None,
```

Is there a case where it makes sense for `perturbation` to be `None`? I think it is fine to have it as the default value, because you added this as the last argument. However I wonder if this should raise an error if `perturbation` is left `None`. Or could it be that a user would pass the list of perturbations directly in the input set generator?

---

I let it as it is for now, a user could indeed pass the perturbation directly to the input set generator via a power up for example. Maybe there is a way to check it before the abipy factory throws an Error ?